Caching files

mjsobrep opened the pull request:
*Issue:* #19 

*Description of changes:*
Added support for caching files by saving them and keeping note of them in sqlite db. Cache file size is tracked and when a maximum is exceeded, the least recently used file is deleted. 

*TODO*

- [ ] Expose the cache size to the launch system
- [ ] Allow the cache location to be set in the launch system
- [ ] Test more widely (only tested with the simple action server example)
- [ ] Document all of this for general use
- [ ] Add unit tests for this functionality

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mjsobrep committed:
Changed hash to work on the entire request contents

mjsobrep committed:
Got a reasonable version of cached audio working

mjsobrep committed:
changed to using integers for byte size

mjsobrep committed:
made the cache size a parameter

This still needs to be added to the documentation and exposed to the roslaunch system

mjsobrep committed:
added logging and clarified a variable name

mjsobrep committed:
formatting fixes

mjsobrep committed:
made cache only run if no file is specified

mjsobrep commented:
I can't get the code coverage up because of #37 . When those tests do start running, I expect that it will be a good idea to add random strings to the end of any test text that need to be pinging the server. Otherwise repeated runs of tests will just hit the cache. There also need to be some tests which intentionally hit the cache.

mjsobrep commented:
@AAlon is this something that you all are interested in? Should I keep working on it?

AAlon commented:
> @AAlon is this something that you all are interested in? Should I keep working on it?

Definitely! thank you for contributing. We'll be reviewing this soon.

cevans87 reviewed (changes requested):
Thanks for submitting this. I'm a fan of the approach. Just a few things that I think could show up as bugs and some suggestions.

I'll get back to you on how we can deal with test coverage after I understand #37 a bit better.

mjsobrep committed:
fixed spelling in DB class docstring

mjsobrep committed:
re-raised exception if database cannot be connected to

mjsobrep commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,82 @@
+
+        try:
+            self.conn = sqlite3.connect(db_location)
+        except sqlite3.OperationalError as e:
```
Yeah, it should re-raise the exception.

mjsobrep committed:
removed unused con function in the db class

mjsobrep committed:
added a function to get the size of the cache

mjsobrep committed:
added a function to get the number of files in the cache

mjsobrep committed:
added a function to remove files from cache

mjsobrep committed:
removed the size table

mjsobrep committed:
changed to not use size tables in synthesizer. Made more robust to errors

mjsobrep commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class DB(object):
+    """DB a class to manage the database for tracking chanced files"""
```
fixed spelling

mjsobrep commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,82 @@
+                to_return = self.conn.execute(command)
+        return to_return
+
+    def con(self):
```
haha. True, not sure why that was still around. Removed.

mjsobrep commented on tts/src/tts/synthesizer.py:
```diff
@@ -103,22 +108,80 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
+            if db_search_result:  # then there is data
+                db.ex('update  cache set last_accessed=? where hash=?',
+                      current_time, tmp_filename)
+                synth_result = PollyResponse(json.dumps({
```
Good catch. Done. I also added in a bit to delete the file from the database if the file does not exist on disk

mjsobrep commented on tts/src/tts/synthesizer.py:
```diff
@@ -103,22 +108,80 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
+                        remove_res = db.ex(
+                            'select hash, file, min(last_accessed), size from cache'
+                        ).fetchone()
+                        os.remove(remove_res['file'])
```
pulled this out into this function which checks that the file exists: https://github.com/aws-robotics/tts-ros1/pull/36/files#diff-a95048b06eab38a802435dada9d4ea64R72

mjsobrep committed:
made sure that the get size function returns a 0 if no rows are present

mjsobrep committed:
made sure that the get size function returns a 0 if no rows are present

mjsobrep commented on tts/src/tts/synthesizer.py:
```diff
@@ -103,22 +108,80 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
+                    rospy.loginfo(
+                        'generated new file, saved to %s and cached', file_name)
+                    # make sure the cache hasn't grown too big
+                    while total_size > self.max_cache_bytes and num_files > 1:
```
Huh, seems cool to me. I did something similar but pulled the size out as a function in the db class. If someone were to use a massive cache, this could slow everything down, but the rest of the codebase isn't really fast anyway...

mjsobrep committed:
got the right return values in get size and get num files

mjsobrep committed:
Merge branch 'master' of https://github.com/mjsobrep/tts-ros1

mjsobrep commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,82 @@
+            size integer NOT NULL
+            );''')
+
+        self.ex('''CREATE TABLE IF NOT EXISTS size (
```
Yeah, I think you are right. I changed it to work like this.

mjsobrep commented on tts/src/tts/synthesizer.py:
```diff
@@ -103,22 +108,80 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
+                        num_files = num_files - 1
+                        rospy.loginfo('removing %s to maintain cache size, new size: %i',
+                                      remove_res['file'], total_size)
+                    db.ex(
```
size table is gone

mjsobrep committed:
fixed error to pull the file name when checking on disk

mjsobrep committed:
added a warning if a file is found in the db and not on disk

mjsobrep commented:
@cevans87 Thanks for the feedback, sorry for being slow to get back on it. I have gotten rid of the size table per your recommendation and pulled some stuff in the db class to improve readability. I have tested under both normal operation and deleting cached files from the disk without telling the db.

mjsobrep commented:
I just found a bug with this: If there is no network connection, then the error utterance (' the polly service cannot be reached ...') gets cached as whatever is being said. Then the system will always say that when the text is being said.

AAlon commented:
> I just found a bug with this: If there is no network connection, then the error utterance (' the polly service cannot be reached ...') gets cached as whatever is being said. Then the system will always say that when the text is being said.

Are you actively working on that? if so we should probably wait before re-reviewing. It'd be great if you could add some unit tests too. Thanks!

mjsobrep commented:
@AAlon I took a look at it a while back, but haven't had time to come up with a good fix. It seems like the way that errors are handled is a little odd. I will take a look at fixing it eventually, but it will require the errors to propagate up rather than just being handled by passing a different sound file. I actually think that would be a good thing. In many situations the current error handling would be poor for human robot interaction. Allowing integrators (like me) to customize how errors are handled would be better. I should get to that at the latest by the middle of February. I'm pretty busy until then. 

I'm not sure how to add unit tests, ref #37

mjsobrep committed:
do not cache files if there is an exception

mjsobrep committed:
more tests for the caching

mjsobrep committed:
fixed db import in tests

mjsobrep committed:
fixed metadata in tests

mjsobrep committed:
fixed problems with the dummy engine

mjsobrep committed:
added a bit more logging

mjsobrep committed:
added a test for audio files going missing

mjsobrep committed:
added a test for removing files from db as it grows

mjsobrep committed:
added more tests for cleaning up files

mjsobrep committed:
added test to make sure that caching works when there is no connection

mjsobrep commented:
@cevans87 @AAlon I fixed the caching of the error file and added a bunch of new tests to get the code coverage up to a reasonable level. I think this is ready for a new review.

mjsobrep commented:
Want to check back in on this.

mm318 commented on tts/src/tts/synthesizer.py:
```diff
@@ -83,15 +87,82 @@ def __call__(self, **kwargs):
+            Args:
+                size: the number of bytes to make the next files
+            """
+            self.file_size = size
```
I'm kind of on the fence about this class being here instead of being under `test`, such as in `test_unit_synthesizer.py`.

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+        import uuid
+
+        db = DB()
+        init_num_files = db.get_num_files()
```
Should you have an assert that `init_num_files == 0`? So later on, your test should be `db.get_num_files() == 1` instead of `db.get_num_files() == init_num_files + 1`.

mm318 replied:
This may need a deletion of `/tmp/polly.db` at the end of each test case.

What is the right way to delete a database and all the audio files associated with it?

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+            self.assertEqual(db.get_num_files(), init_num_files + 1)
+
+        
+    def test_multiple_novel(self):
```
It's probably better named `test_multiple_unique`.

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+        self.assertEqual(db.get_num_files(), init_num_files)
+
+    def test_no_connection_existing(self):
+        from tts.db import DB
```
Unused import.

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+        import json
+        import os
+
+        db = DB()
```
Unused variable.

mm318 replied:
Though, it would be better to have some tests for `db.get_num_files()`.

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+        self.assertEqual(audio_file1, audio_file2)
+        self.assertTrue(os.path.exists(audio_file2))
+
+    def test_no_connection_novel(self):
```
`test_no_connection_nonexisting`?

mm318 commented on tts/test/test_unit_synthesizer.py:
```diff
@@ -142,6 +142,200 @@ def test_cli_engine_dispatching_3(self, speech_synthesizer_class_mock):
+
+        self.assertFalse(os.path.exists(audio_file1))
+
+    def test_big_db(self):
```
I think `test_db_size_limit` would be a better name.

mm318 reviewed (commented):
Thank you very much for the significant contribution!

I have some comments and questions.

AAlon commented on tts/src/tts/synthesizer.py:
```diff
@@ -103,22 +174,81 @@ def __init__(self, engine='POLLY_SERVICE', polly_service_name='polly'):
+            file_found = False
+            if db_search_result:  # then there is data
+                # check if the file exists, if not, remove from db
+                # TODO: add a test that deletes a file without telling the db and tries to synthesize it
+                if os.path.exists(db_search_result['file']):
+                    file_found = True
+                    db.ex('update  cache set last_accessed=? where hash=?',
+                          current_time, tmp_filename)
+                    synth_result = PollyResponse(json.dumps({
+                        'Audio File': db_search_result['file'],
+                        'Audio Type': db_search_result['audio_type'],
+                        'Amazon Polly Response Metadata': ''
+                    }))
+                    rospy.loginfo('audio file was already cached at: %s',
+                                  db_search_result['file'])
+                else:
+                    rospy.logwarn(
+                        'A file in the database did not exist on the disk, removing from db')
+                    db.remove_file(db_search_result['file'])
+            if not file_found:  # havent cached this yet
+                rospy.loginfo('Caching file')
+                synth_result = self.engine(**kw)
+                res_dict = json.loads(synth_result.result)
+                if 'Exception' not in res_dict:
+                    file_name = res_dict['Audio File']
+                    if file_name:
+                        file_size = os.path.getsize(file_name)
+                        db.ex('''insert into cache(
+                            hash, file, audio_type, last_accessed,size)
+                            values (?,?,?,?,?)''', tmp_filename, file_name,
+                              res_dict['Audio Type'], current_time, file_size)
+                        rospy.loginfo(
+                            'generated new file, saved to %s and cached', file_name)
+                        # make sure the cache hasn't grown too big
```
#### Performance concerns
* Potentially doing many more transactions than needed; only need to do at most one transaction per _call_engine call. For example, if the new file caused us to go beyond the size limit by 1000 bytes, and we need to remove the 10 oldest files in order to clear that much space, we'd be doing more than 10 transactions.
* Seems like everything is synchronous, so even if we only have one transaction (say, to insert a new entry) - returning `synth_result` is still blocked on completion of that transaction. Using WAL and performing checkpointing in a separate thread is one way to overcome that.

#### Performance Impact discussion
In part, the value of this node is in enabling Human-Robot interaction via Polly, and that heavily depends on the latency; if I'm asking the robot a question and it answers 5 seconds later, that's not very useful. 

We know that with this implementation there will be increased latency, the question is how much. Would it add 1ms, or 500ms? maybe it's even less than 1ms, I really don't know. Do you? 
It would feel a bit premature to me to merge this without doing a simple benchmark in which we mock out line 228 (the self.engine() call) and measure the added latency (in lieu of any other data).

#### Path forward
Generally, it seems like we're reinventing the wheel here by trying to use SQLite directly. There are dozens of libraries that implemented all the convenience functionality we'd ever need (see [DiskCache](http://www.grantjenks.com/docs/diskcache/tutorial.html#cache)) as well as performance / transaction optimizations. So to summarize:
* Why not use some of the battle tested disk-backed caches out there? Did you choose sqlite because of lack of integration with ROS? (i.e. not available via rosdep?)
* If we have to continue using SQLite directly, we need to test the effect on latency. Again, maybe there's no issue at all, but we need to make sure.

mjsobrep replied:
Sorry I have been so slow to respond. I'm going to start here before going anywhere else, since the answers on this thread may change what is being done elsewhere.

I totally agree that latency is a major issue in these types of systems. SQLite is fast and I would imagine that it is at least an order of magnitude faster than making a web call that downloads a file, probably multiple orders of magnitude better. But you are correct, I don't know that for a fact. 

I chose SQLite because it is built into python, it is stable, and it is well known/documented. I could certainly see another tool being used, but I don't know which other tools would check those boxes. DiskCache that you link to looks great. Is that a dependency you all are ok with?

mjsobrep replied:
Hey @AAlon, want to check back in on this. If you would like me to refactor this to DiskCache or make some timing tests for the current implementation, I can do either of those. DiskCache would make for cleaner more maintainable code, but not sure if that is a dependency that y'all want in here. Just let me know, I'm hoping to get some code done in a few places over the next few weeks.

dabonnie replied:
I think that's fine, but we should probably take a look at the benchmarking mentioned previously.

mjsobrep replied:
what is "that" in "that's fine"?

dabonnie replied:
Sorry, I'll be specific. @mjsobrep if you would like to implement some timing tests, in order to provide a ground truth, and test your changes here (along with possibly refactoring to DiskCache) I think that's a good start. 

I'm not sure if you'e seen it, but the [DiskCache timing info](http://www.grantjenks.com/docs/diskcache/#comparisons) seems promising.

AAlon commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,97 @@
+class DB(object):
+    """DB a class to manage the database for tracking cached files"""
+
+    def __init__(self, db_location='/tmp/polly.db'):
```
tempfile.gettempdir() instead of /tmp?

AAlon commented on tts/src/tts/db.py:
```diff
@@ -0,0 +1,97 @@
+    def make_db(self):
+        self.ex('''CREATE TABLE IF NOT EXISTS cache (
+            hash text PRIMARY KEY,
+            file text NOT NULL,
```
`file` is actually a `filepath`, correct?

AAlon reviewed (commented):
Thanks again for taking the time and effort to iterate on this @mjsobrep! Most of my comments are very minor, but I do have one concern around performance and SQLite I'd like to have addressed before finalizing on this implementation.

mjsobrep committed:
Merge branch 'master' of https://github.com/aws-robotics/tts-ros1
